### ISSUE TITLE
`willRenameFile` request on `save-buffer` sends backup file URI

### ISSUE THREAD

**Issue by Walheimat** [emacs-lsp/lsp-mode#3516]
### Thank you for the bug report

- [X] I am using the latest version of `lsp-mode` related packages.
- [X] I checked [FAQ](https://emacs-lsp.github.io/lsp-mode/page/faq/) and [Troubleshooting](https://emacs-lsp.github.io/lsp-mode/page/troubleshooting/) sections
- [ ] You may also try reproduce the issue using clean environment using the following command: `M-x lsp-start-plain`

### Bug description

When editing files using a language server that supports file renaming, saving a file sends a `willRenameFiles` request providing a new URI of the **backup** file. It seems that this is because over the course of `save-buffer` a rename happens by default.

I found this in the docs of `make-backup-files`:

> Non-nil means make a backup of a file the first time it is saved.
> This can be done by renaming the file or by copying.
> Renaming means that Emacs renames the existing file so that it is a
> backup file, then writes the buffer into a new file.  Any other names
> that the old file had will now refer to the backup file.  The new file
> is owned by you and its group is defaulted.

### Steps to reproduce

You will need a LSP server that supports `willRename` file operation. The only one I use is the Dart Analyzer via `lsp-dart`.

In the case of Dart, importing a file in another file and changing that file and saving it will update the import and make it point to the backup file.

### Expected behavior

No such request is sent as no renaming took place outside of Emacs doing backups.

### Which Language Server did you use?

lsp-dart (which is an external project I know but the only one I use that supports this).

### OS

Linux

**Comment by Walheimat** [comment]
Yes, I'm currently trying to reproduce this on a different machine/project.

**Comment by Walheimat** [comment]
Okay, I can recreate this the following way (using Flutter):

1. Create a project:
    ```bash
     flutter create example
    ```
2. Open `lib/main.dart` and import root folder when prompted.
3. Edit the file (empty line for example) and save.
4. Test file `<root>/test/widget_test.dart` now imports:
    ```dart
    import '../../../../.emacs.d/backups/!home!karusster!tests!flutters!nexample!lib!main.dart~';
    ```

So it looks like this happens on first save, when the backup file is created. I assume this can be circumvented by setting `backup-by-copying` to `t` but will have to test first.

**Comment by ericdallo** [comment]
@Walheimat that's weird, when did you rename a file so lsp-mode would send `willRenameFiles` request?

**Comment by ericdallo** [comment]
BTW I can't repro on doom-emacs

**Comment by Walheimat** [comment]
I will just set `backup-by-copying` myself but other vanilla users might run into this issue. Maybe `lsp-mode` could warn if the server supports file renaming but the variable is `nil`?

**Comment by yyoncho** [comment]
> I will just set `backup-by-copying` myself but other vanilla users might run into this issue. Maybe `lsp-mode` could warn if the server supports file renaming but the variable is `nil`?

Something like that and then suggest either disabling lsp-mode feature or changing the backup setting.

**Comment by apergos** [comment]
I just got bit by this bug, tracked it down to the backup files and willRenameFiles, vanilla emacs 29.1, lsp-mode 20231203.705 and I disabled backups of files with relevant extensions to work around it, since those files will end up in a git repo anyways. Only then did I find this open issue. In my case, the language server threw an exception because it didn't like the rename, which threw me off the track.

**Comment by slipset** [comment]
I experienced this bug as well when setting up with a new emacs config.
This config did some custom stuff with backups:
```
(setq backup-directory-alist
      `(("." . ,(expand-file-name
                 (concat user-emacs-directory "backups")))))
```
Which made lsp-mode send rename requests on the first save, which then created havoc in my project :)
I have now commented out the backup stuff, and lsp-mode does not create havoc.

**Comment by justinbarclay** [comment]
I get bit by this bug so often.

**Comment by Willyfrog** [comment]
took me a while to find this thread, but same issue.
My problem is when creating a new file it will create a backup copy and seems it is trying to rename it and the LSP is rejecting the file as it is not part of the configured extensions for renaming.

repro steps:
- go to project
- let LSP load
- have an emacs backup file (e.g.: `#theme.ts#`) from a previous session, but never got around to saving it (this happens with new files if you take long enough until save)
- create a new file that would match that backup file (in our example: `theme.ts`) and write some code into it.
- try to save

output:
timeout on willRename, and the file never gets saved

expected:
file saves, no timeout

attached a picture with the filter that seems to be applying to the willSave function. I've been trying to figure out how to add a regex for backup files so this doesn't happen.
![Image](https://github.com/user-attachments/assets/8e1a3c3c-431f-4c09-a5f8-7a6819b8362a)

I'll try the backup-by-copying option in the meantime

**Comment by Konubinix** [comment]
I just came across this issue, also did my digging and came to that workaround

```
(defun konix/inhibit-lsp-during-backup-buffer (orig-fn &rest args)
  "Inhibit lsp-mode file notifications when running backup-buffer."
  (let ((lsp-apply-edits-after-file-operations nil))
    (apply orig-fn args)))

(advice-add 'backup-buffer :around #'konix/inhibit-lsp-during-backup-buffer)
```

Now that I see the backup-by-copying setting, I will use it, but I thought that it was worth sharing.
